### DISCUSSION TITLE
Air Quality Systems

### DISCUSSION THREAD

**Discussion by TBN-MapleWheels** [PeasantFoundry/Mechtrauma#46] [Ideas]
Originally suggested by @Torca2001 . Repost from the Wiki.

Game Mechanic: Have multiple gases tracked (O2, CO2, Other) within hull sections and have status effects caused by them. Examples are things like euphoria from hypoxia and carbon monoxide poisoning.

**Comment by PeasantFoundry** [comment]
This would significantly improve the game mechanics for carbon dioxide and carbon monoxide poisoning. Currently, carbon dioxide poisoning is transmitted via an XML explosion that increases a carbon dioxide affliction that slowly goes away on it's own. Currently this explosion only transmits when an oxygen vent is active and has no working co2 filter.

The current system works surprisingly well considering how much of a colossal hack it is. The main problems with it are 
1) the ability for c02 poisoning to be transmitted between hulls that it shouldn't be
2) there is no accounting for c02 production so c02 poisoning always increases at a steady rate (IE same level of poisoning for person vs 10 people in the same hull)
3) there is no accounting for how much co2 in currently in the hull so co2 poisoning always increases at a steady rate
4) c02 filtering cannot be reasonably centralized 

Additional Uses:
1) The ability to track particulates (such as airborne spores)
2) The ability to transfer gasses and particulates into oxygen tanks (tank compressors running in contaminated environments could create contaminated oxygen bottles) yes - getting co2 poisoning or a spore infection from a cursed o2 tank.

**Comment by MapleWheels** [comment]
I have already partially made a solution to this to track custom values on my fork `air-quality-systems`. What I'm looking into is a clean way to run an update whenever the hull section updates. Give me a list of what to track and I'll add it to the class data.

**Reply by dakkhuza** [reply]
Looked at your branch, you could simplify the data storage a lot by using conditional weak tables, e.g. here's a data storage class I use for More Level Content
```
    public static partial class MLCData
    {
        private static readonly ConditionalWeakTable<Level, Level_MLCData> level_data = new();

        internal static Level_MLCData MLC(this Level levelData) => level_data.GetOrCreateValue(levelData);

        internal static void AddData(this Level levelData, Level_MLCData additional)
        {
            try
            {
                level_data.Add(levelData, additional);
            }
            catch (Exception e) { Log.Error(e.ToString()); }
        }
    }
```
Lets me just do `levelDataInstance.MLC()` and instantly get access to any data I want to attach to the instance, the bonus is references don't persist so you don't have to manage removing hulls from your internal lists when they get deleted by the game

**Reply by MapleWheels** [reply]
I actually forgot about `ConditionalWeakTable<>`. Big thanks for the reminder.
